encode.js

```diff
@@ -604,32 +604,157 @@ modeSelect.addEventListener("change", (e) => {
+    for (i = 0; i < abuffer.numberOfChannels; i++)
+        channels.push(abuffer.getChannelData(i));
+
+    while (pos < length) {
+        for (i = 0; i < numOfChan; i++) {
+            // interleave channels
+            sample = Math.max(-1, Math.min(1, channels[i][offset])); // clamp
+            sample = (0.5 + sample * 0.5) * 0xffff; // scale to 16-bit unsigned
+            sample = Math.floor(sample);
+            view.setUint16(pos, sample, true); // write 16-bit sample
+            pos += 2;
+        }
+        offset++; // next source sample
+    }
 
-	if(modeSelect.value == "none") {
-		warningText.textContent = "You must select a mode";
-		startButton.disabled = true;
-		return;
-	}
+    return buffer;
 
-	if(!imageLoaded){
-		warningText.textContent = "You must upload an image";
-		startButton.disabled = true;
-		return;
-	}
+    function setUint16(data) {
+        view.setUint16(pos, data, true);
+        pos += 2;
+    }
+
+    function setUint32(data) {
+        view.setUint32(pos, data, true);
+        pos += 4;
+    }
+}
+
+function saveWavFile(buffer, filename) {
+    var blob = new Blob([buffer], { type: 'audio/wav' });
+    var url = window.URL.createObjectURL(blob);
+    var a = document.createElement('a');
+    a.style.display = 'none';
+    a.href = url;
+    a.download = filename;
+    document.body.appendChild(a);
+    a.click();
+    window.URL.revokeObjectURL(url);
+}
 
-	let canvasData = canvasCtx.getImageData(0, 0, canvas.width, canvas.height);
+let saveButton = document.createElement("button");
+saveButton.id = "saveButton";
+saveButton.className = "contrast";
+saveButton.textContent = "Save as WAV";
+saveButton.disabled = true;
+document.querySelector("main.container").appendChild(saveButton);
+
+startButton.onclick = () => {
+    if (modeSelect.value == "none") {
+        warningText.textContent = "You must select a mode";
+        startButton.disabled = true;
+        return;
+    }
+
+    if (!imageLoaded) {
+        warningText.textContent = "You must upload an image";
+        startButton.disabled = true;
+        return;
+    }
+
+    let canvasData = canvasCtx.getImageData(0, 0, canvas.width, canvas.height);
 
 	warningText.textContent = "";
 	if (audioCtx.state === "suspended") {
-    	audioCtx.resume();
-    }
+		audioCtx.resume();
+	}
+
+	sstvFormat.getDuration = function() {
+		return (this.numScanLines * (this.scanLineLength + this.blankingInterval + this.syncPulseLength * 3);
+	};
 
-    let oscillator = audioCtx.createOscillator();
+	let oscillator = audioCtx.createOscillator();
 	oscillator.type = "sine";
 
+	let dest = audioCtx.createMediaStreamDestination();
+	oscillator.connect(dest);
 	oscillator.connect(audioCtx.destination);
 
 	sstvFormat.prepareImage(canvasData.data);
 	sstvFormat.encodeSSTV(oscillator, audioCtx.currentTime + 1);
-};
\ No newline at end of file
+
+	let recorder = new MediaRecorder(dest.stream);
+	let chunks = [];
+
+	recorder.ondataavailable = (e) => {
```

Comment:
Instead of using the `MediaRecorder` API, I'd prefer if you used the `OfflineAudioContext` api.